tasks/ast_codegen/src/main.rs

```diff
@@ -0,0 +1,330 @@
+use std::{borrow::Cow, collections::HashMap, error::Error, io::Read, path::PathBuf};
+
+use itertools::Itertools;
+use syn::{parse_file, Ident};
+
+#[derive(Default)]
+struct AstCodegen {
+    files: Vec<PathBuf>,
+}
+
+type Result<R> = std::result::Result<R, String>;
+
+impl AstCodegen {
+    #[must_use]
+    fn add_file<P>(mut self, path: P) -> Self
+    where
+        P: AsRef<str>,
+    {
+        self.files.push(path.as_ref().into());
+        self
+    }
+
+    fn generate(self) -> Result<String> {
+        let mut schema = self
+            .files
+            .into_iter()
+            .map(Module::from)
+            .flat_map(Module::load)
+            .map(Module::expand)
+            .collect::<Result<Vec<_>>>()?;
+        let mut ty_table: HashMap<String, &mut AstType> = schema
+            .iter_mut()
+            .flat_map(|it| it.items.iter_mut())
+            .map(|it| (it.ident().to_string(), it))
+            .collect();
+        for (ident, ty) in &mut ty_table {
+            dbg!(ident, ty);
+        }
+        // dbg!(file_mod.keys());
+        // .map_ok(Module::build)
+        // .flatten()
+        // .collect();
+
+        // for s in schema.definitions {}
+
+        Ok(String::new())
+    }
+}
+
+use analyzer::*;
+use schema::*;
+mod schema {
+    use std::borrow::Cow;
+
+    use syn::{
+        braced,
+        parse::{Parse, ParseBuffer},
+        punctuated::Punctuated,
+        token, Attribute, File, Generics, Ident, Item, ItemEnum, ItemMacro, ItemStruct, Token,
+        Variant, Visibility,
+    };
+
+    use super::*;
+
+    #[derive(Debug)]
+    pub struct Schema {
+        source: PathBuf,
+        definitions: Definitions,
+    }
+
+    #[derive(Debug)]
+    pub enum Inherit {
+        Unlinked(Ident),
+    }
+
+    impl From<Ident> for Inherit {
+        fn from(ident: Ident) -> Self {
+            Self::Unlinked(ident)
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct ItemEnumExt {
+        pub attrs: Vec<Attribute>,
+        pub vis: Visibility,
+        pub enum_token: Token![enum],
+        pub ident: Ident,
+        pub generics: Generics,
+        pub brace_token: token::Brace,
+        pub variants: Punctuated<Variant, Token![,]>,
+        pub inherits: Vec<Inherit>,
+    }
+
+    impl From<ItemEnum> for ItemEnumExt {
+        fn from(other: ItemEnum) -> Self {
+            Self {
+                attrs: other.attrs,
+                vis: other.vis,
+                enum_token: other.enum_token,
+                ident: other.ident,
+                generics: other.generics,
+                brace_token: other.brace_token,
+                variants: other.variants,
+                inherits: Vec::new(),
+            }
+        }
+    }
+
+    #[derive(Debug, Default)]
+    pub struct EnumMeta {
+        inherits: Vec<Inherit>,
+    }
+
+    #[derive(Debug)]
+    pub struct AstEnum {
+        pub tt: ItemEnum,
```

Comment:
nit: I try to avoid abbreviations. I will never know what this stands for the first time.

Reply:
You are right, I'll change them in the cleanup phase. I just named stuff similar to Rust's macro convention.